services/web/server/src/simcore_service_webserver/projects/_projects_repository_legacy.py

```diff
@@ -996,14 +919,103 @@ async def update_project_multiple_node_data(
             msg=f"update multiple nodes on {project_uuid=} for {user_id=}",
             extra=get_log_record_extra(user_id=user_id),
         ):
-            return await self._update_project_workbench(
+            return await self._update_project_workbench_with_lock_and_notify(
                 partial_workbench_data,
                 user_id=user_id,
-                project_uuid=f"{project_uuid}",
+                project_uuid=project_uuid,
                 product_name=product_name,
                 allow_workbench_changes=False,
             )
 
+    async def _update_project_workbench_with_lock_and_notify(
+        self,
+        partial_workbench_data: dict[NodeIDStr, Any],
+        *,
+        user_id: UserID,
+        project_uuid: ProjectID,
+        product_name: str | None = None,
+        allow_workbench_changes: bool,
+    ) -> tuple[ProjectDict, dict[NodeIDStr, Any]]:
+        """
+        Updates project workbench with Redis lock and user notification.
+
+        This method performs the following operations atomically:
+        1. Updates the project workbench in the database
+        2. Retrieves the updated project with workbench
+        3. Creates a project document
+        4. Increments the document version
+        5. Notifies users about the project update
+
+        Note:
+            This function is decorated with Redis exclusive lock to ensure
+            thread-safe operations on the project document.
+        """
+
+        @exclusive(
+            get_redis_lock_manager_client_sdk(self._app),
+            lock_key=PROJECT_DB_UPDATE_REDIS_LOCK_KEY.format(project_uuid),
```

Comment:
so as I said earlier:
- I would suggest to create either a context manager or another decorator that is living in your project_document .py file
- --> you do not need to expose the keys directly and they remain private and you fully control in one module how the keys are used.
- --> you ensure that the blocking and blockingTimeout are the same everywhere